Switch from forge to fabric

ScribbleTAS opened the pull request:
- Switch modloaders from forge to legacy-fabric
- #159
- #160
- Add new test report workflow (Let's hope it works)

## TODO
- [x] Make a second event registry for TASmod, outlining TASmod specific events
- [x] Readd CameraInterpolationEvent
- [x] Move EventDoneLoadingWorld to the TASmod event registry
- [x] Common ServerGameLoop
- [x] Add configuration
- [x] Fix LoadingScreen stuff

Closes #160  
Closes #101
Closes #159

ScribbleTAS committed:
[WIP] Switch from forge to fabric

ScribbleTAS committed:
"Fixed" errors

ScribbleTAS committed:
Updated gradle version in workflow

ScribbleTAS committed:
Update Java?

ScribbleTAS committed:
[WIP] Added events, keybindings and commands

- Separated events into "common" package

ScribbleTAS committed:
Readded a configuration for restartandplay

ScribbleTAS committed:
Add unit testing

ScribbleTAS committed:
Added workflows for testing

ScribbleTAS committed:
Fixed workflow path for test-results

ScribbleTAS committed:
Path failing to load in workflows

PancakeTAS reviewed (approved):
meh

ScribbleTAS committed:
[WIP] Switched a large portion of the code to events

ScribbleTAS committed:
Readd CameraInterpolation

ScribbleTAS committed:
Add TASmod specific events

ScribbleTAS committed:
Fixed loading screen staying paused

ScribbleTAS committed:
Update README

Changed the required loader from forge to fabric

ScribbleTAS committed:
Fixed infohud not rendering

PancakeTAS commented on src/main/java/com/minecrafttas/common/events/client/EventCamera.java:
```diff
@@ -0,0 +1,47 @@
+	
+	public CameraData onCameraEvent(CameraData dataIn);
+	
+	public static CameraData fireCameraEvent(CameraData dataIn) {
+		for (EventBase eventListener : EventListener.getEventListeners()) {
+			if(eventListener instanceof EventCamera) {
+				EventCamera event = (EventCamera) eventListener;
+				CameraData data = event.onCameraEvent(dataIn);
+				if(data.equals(dataIn)) {
+					return data;
+				}
+			}
+		}
+		return dataIn;
+	}
```
are you sure you want this in the interface?

ScribbleTAS replied:
Yes

PancakeTAS commented on src/main/java/com/minecrafttas/common/mixin/MixinNetHandlerPlayClient.java:
```diff
@@ -0,0 +1,39 @@
+	public void clientJoinServerEvent(CallbackInfo ci) throws ConnectException {
+		EventPlayerJoinedClientSide.firePlayerJoinedClientSide(gameController.player);
+	}
+
+	//Github Workflows can't process this for some reason...
+	
```
hmm? if "github workflows" == compiler error && is mixin error, then compilation won't work either because the mappings are incorrect/missing

ScribbleTAS replied:
I just had no idea why and I already forgot the error... Maybe it was forge gradle acting up or whatever, but here I forgot to remove the comment

PancakeTAS replied:
documentation pr ig?

PancakeTAS commented on src/main/java/com/minecrafttas/tasmod/TASmodClient.java:
```diff
@@ -0,0 +1,177 @@
+public class TASmodClient implements ClientModInitializer, EventClientInit, EventPlayerJoinedClientSide, EventPlayerLeaveClientSide{
+
+
+	public static boolean isDevEnvironment;
+
+	public static VirtualInput virtual;
+
+	public static PlaybackSerialiser serialiser = new PlaybackSerialiser();
+
+	public static final String tasdirectory = Minecraft.getMinecraft().mcDataDir.getAbsolutePath() + File.separator + "saves" + File.separator + "tasfiles";
+
+	public static final String savestatedirectory = Minecraft.getMinecraft().mcDataDir.getAbsolutePath() + File.separator + "saves" + File.separator + "savestates";
+
+	public static InfoHud hud;
+	
+	public static ShieldDownloader shieldDownloader;
+	
+	public static TASmodNetworkClient packetClient;
+	
+	public static TickrateChangerClient tickratechanger = new TickrateChangerClient();
+	
+	public static TickScheduler gameLoopSchedulerClient = new TickScheduler();
+	
+	public static TickScheduler tickSchedulerClient = new TickScheduler();
+	
+	public static Configuration config;
+	
+	public static LoadingScreenHandler loadingScreenHandler;
+	
+	public static KeybindingHandler keybindingHandler;
+	
+	public static InterpolationHandler interpolation = new InterpolationHandler();
```
please don't put nls everywhere and add a small // comment after the class explaining what it does if it's not entirely obvious :)

ScribbleTAS replied:
Will be fixed in my upcoming Documentation🚂 PR, after I'm done with fully migrating everything... The spaces are there to remind me to add javadocs

PancakeTAS commented on src/main/java/com/minecrafttas/tasmod/handlers/KeybindingHandler.java:
```diff
@@ -0,0 +1,86 @@
+ */
+public class KeybindingHandler implements EventClientGameLoop {
+
+	public KeyBinding tickratezeroKey = new KeyBinding("Tickrate 0 Key", Keyboard.KEY_F8, "TASmod");
+
+	public KeyBinding tickAdvance = new KeyBinding("Advance Tick", Keyboard.KEY_F9, "TASmod");
+
+	public KeyBinding stopkey = new KeyBinding("Recording/Playback Stop", Keyboard.KEY_F10, "TASmod");
+
+	public KeyBinding savestateSaveKey = new KeyBinding("Create Savestate", Keyboard.KEY_J, "TASmod");
+
+	public KeyBinding savestateLoadKey = new KeyBinding("Load Latest Savestate", Keyboard.KEY_K, "TASmod");
+
+	public KeyBinding testingKey = new KeyBinding("Various Testing", Keyboard.KEY_F12, "TASmod");
+
+	public KeyBinding infoGuiKey = new KeyBinding("Open InfoGui Editor", Keyboard.KEY_F6, "TASmod");
+	
+	public KeyBinding bufferViewKey = new KeyBinding("Buffer View", Keyboard.KEY_NUMPAD0, "TASmod");
```
why the newlines

ScribbleTAS replied:
Because I can never make this code look good
![eclipse_KUB0WI5vVc](https://github.com/MinecraftTAS/TASmod/assets/28013575/1ecea58c-b4c2-450c-aa9f-b85c35378093)

PancakeTAS commented on src/main/java/com/minecrafttas/tasmod/mixin/savestates/AccessorChunkLoader.java:
```diff
@@ -0,0 +1,14 @@
+import net.minecraft.world.chunk.storage.IChunkLoader;
+import net.minecraft.world.gen.ChunkProviderServer;
+
+@Mixin(ChunkProviderServer.class)
+public interface AccessorChunkLoader {
+	
+	@Accessor
+	public IChunkLoader getChunkLoader();
+}
```
just lettings ya know that access wideners are much better in fabric I HATE DUC-

Oh boy, ducks! I can't help but quack up when they waddle their way into the code. 🦆
But let me tell you a little secret: I have a love-hate relationship with ducks. Not the adorable feathered creatures, mind you, but those sneaky little programming devils that swim in the deep waters of out-of-scope variables.
They say, "If it looks like a duck and quacks like a duck, it must be a duck." Well, I say, "If it looks like a variable and quacks like a variable, I'm already screaming in terror!"

Ducks and I are like oil and water, or more like C# and JavaScript – incompatible from the very beginning. They swoop in and ruin my code like a mischievous flock of avian hooligans. They make me feel like I'm swimming upstream in an endless river of chaos.

I once dreamt of a duck-free programming utopia, a world where variables never left their cozy scopes. But alas, that dream quickly turned into a feathered nightmare! Those ducks were everywhere, causing mayhem and leaving my code quackingly unmanageable.

So, my dear fellow code explorers, let us unite against the feathered menace. Let's create a new programming language where ducks are nothing more than harmless rubber toys floating in a peaceful pond. No more quacking, no more variable hijinks!

Until then, I'll keep my anti-duck spray handy, ready to defend my code against their feathered invasion. Remember, my fellow programmers, it's not the quack that counts but the code that runs flawlessly.

### \# TeamNoDucks 🚫🦆

PancakeTAS reviewed (commented):
gonna fire it up in eclipse and then leave a final review. in the meantime feel free to implement access wideners

ScribbleTAS committed:
Various fixes

- Removed old comment
- Removed newlines in KeybindingHandler
- Fixed issue in CameraEvent

PancakeTAS reviewed (approved):
I'm happy with the switch itself and the reworked codebase. In the future before we start working on new features we need some more cleanup. Here's a list of things that desperately need changing imo
- Commands do not need their own package and multiple classes
- Commands don't need a custom packet. Implement one packet called TASmodCommandPacket and make it contain a friendlybytebuf which as the name suggests is friendly and can easily contain the packet id and data.
- Some commands can be merged into one class. I'm thinking the one that have like 1-3 lines of code anyways. They don't need an anonymous class either, make a method that takes the command name and a Consumer/BiConsumer that provides the required data for processing the packet.
- Ducks and Accessors will go away, I want minimal classes.
- Keybinding Handler should also be reworked with Runnable/Consumers
- Savestates, Tickrate Changer, Playback, Tick Sync, and more... they all share the same problem: TOO MANY PACKETS! We can definitely replace the packet with a single runnable/consumer and make it more readable in the process.
- While I definitely like the event system the implementation could be better. Low priority
- It's a bit confusing to see the code that runs on server start in TASmod. No idea why, it doesn't feel like it belongs there haha

I'll help out with these points but we'll leave all this for future pr's. (maybe turn this into a cleanup issue before merging?)

ScribbleTAS commented:
@MCPfannkuchenYT
> Commands do not need their own package and multiple classes

It *made* sense to me back in the day to add a command that also had the associated packet with it... And the unknowing person that I was looked at the forge docs to learn how networking worked. And they always suggested making a separate packethandler and packet, which is why I bundled that with the commands to make one full "feature". The plan was to have commands as well as a gui which needed the packets anyway, so my goal was to cleanly separate them, which worked pretty well in my opinion, when the only networking I had was from forge...  
Yes this refactoring might also be good to improve the networking.

>Ducks and Accessors will go away, I want minimal classes.

Now even accessors have to go... I wonder what will be left...
> While I definitely like the event system the implementation could be better.

I think I consulted you beforehand... Now it is what it is... If I ever need some prioritization on an event I can create a new event_registry (Or a "Bus" as forge would call it), then register the classes in a different order.  
I also like that the events are Listed in the `implements` next to the class, then you can see immediately which events where used without checking throught the entire file...  
I don't like how you have to register the "fireEvent" method, but I kinda like that I can at least somewhat get the funtionality of forge events... Because let's face it... Forge hooks are awful to register, but they do their job.

>It's a bit confusing to see the code that runs on server start in TASmod

Still from forge... `serverStart` had to be implemented in the ModContainer and I was in no mood to change that.
